Initial Implementation of Alarms

mrrosen opened the pull request:
Implement alarms with built suppression and trigger for audio

Addresses #4 #15 #16

Commit:
watchdog: implement watchdog timer hal

Implement Watchdog timer HAL to reset the processor after a
specified time for both the Arduino UNO and Arduino Due

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
alarm: add initial alarm infrastructure implementation WIP

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

zmk5 commented on src/hal/alarm.cpp:
```diff
@@ -0,0 +1,76 @@
+  uint32_t alarmTime;
+
+  // Check to see if new tone needs to be played
+  if (pattern != currentPattern) {
```
Should we check that `pattern` is either 0-4?

mrrosen replied:
Done in latest commits

zmk5 commented on src/hal/alarm.cpp:
```diff
@@ -0,0 +1,76 @@
+  
+  if (alarmOn) {
+    // If a blinking buzzer, check the running timer
+    if ((currentPattern == ALARM_HAL_0_25HZ) ||
```
Can we replace this if statement with one that just checks that `currentPattern` isn't `ALARM_HAL_OFF` or `ALARM_HAL_CONSTANT`?

zmk5 replied:
or just `ALARM_HAL_CONSTANT` because looking through the code, it seems it will only hit this section if `currentPattern` is set to `ALARM_HAL_CONSTANT`.

mrrosen replied:
Indeed we can

mrrosen replied:
Done in latest commits

zmk5 commented on src/hal/alarm.cpp:
```diff
@@ -0,0 +1,76 @@
+    }
+    
+    // If a blinking buzzer, check the running timer
+    if ((currentPattern == ALARM_HAL_0_25HZ) ||
```
Same with this one maybe?

zmk5 commented on src/hal/watchdog.cpp:
```diff
@@ -0,0 +1,66 @@
+int watchdogHalInit(void)
+{
+#if defined(__AVR__)
+  if (WATCHDOG_PERIOD_MS == 15) {
```
Maybe replace these with a `switch` statement?

mrrosen replied:
I had a reason I didnt use a `switch` here, but now I dont remember what it was; so yeah, lets use a `switch`

mrrosen replied:
Done in latest commits

zmk5 commented on src/main.cpp:
```diff
@@ -45,5 +50,30 @@ void mainLoop(void)
-  // TODO: Failure actions (alarm, watchdog)
+  // Scan through all alarms, looking for any set warning that need to be addressed
+  struct alarmProperties properties = {0};
+  if (alarmCheckAll(&properties)) {
```
Does the alarm need to be initialized before checked?

mrrosen replied:
Any alarm that has been initialized via `alarmInit` will be checked by this function and the aggregate alarm properties that we need to deal with will be loaded into `properties`; so this isnt checking a specific alarm

zmk5 commented on src/main.cpp:
```diff
@@ -45,5 +50,30 @@ void mainLoop(void)
 
-  // TODO: Failure actions (alarm, watchdog)
+  // Scan through all alarms, looking for any set warning that need to be addressed
+  struct alarmProperties properties = {0};
```
do `preventWatchdog` and `suppressionTimeout` not need initialized values?

mrrosen replied:
`{0}` will zero initialize the entire structure, which is fine for everything we want to use this structure for (getting the properties of the highest priority alarm and whether or not any alarms are triggered that will block the watchdog)

zmk5 commented on src/util/alarm.cpp:
```diff
@@ -0,0 +1,93 @@
+
+  // Loop through all registered alarms, checking if any changes need to be made
+  // to their state, setting the return flags if any conditions are met
+  while (current) {
```
Once the `alarmList` finishes, how does the while look terminate?

mrrosen replied:
Once `current` becomes `NULL`?

zmk5 replied:
Oh yes, the the list should point to a `NULL` at the end.

zmk5 commented on src/util/alarm.cpp:
```diff
@@ -0,0 +1,93 @@
+  bool allClear = true;
+  
+  // Set the priority to below the "lowest" possible value
+  properties->priority == ALARM_PRIORITY_LOW + 1;
```
Where does this get set other than here? Also this line has two equal signs, so it would be a `bool` right?

mrrosen replied:
Thanks for catching the typo. Since I havent added the alarms to the other module yet, I guess it might not be clear how it works. Alarms are initialized via `alarmInit` in which they pass in a struct containing their parameters, like their priority, whether or not they need to prevent the watchdog from being reset, etc. The properties struct here is to aggregate the highest priority settings across all alarms for the main loop to manage.

Commit:
alarm: cleanup audiable alarm hal code

. Add check for valid audio pattern
. Simplify play logic

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
watchdog: clean up watchdog hal

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
alarm: fix bug in alarm checking

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

mrrosen commented:
@zmk5 I have fixed some of the issues/comments you had; please review the changes to make sure I addressed everything

Commit:
alarm: fix mismatch with alarmSuppress

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
main: add some hal initialization during setup

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

mrrosen commented:
@zmk5 I wanted to try it out a little and noticed a few mistakes, can you look over them quick and approve so I cam merge in?

zmk5 commented on src/hal/watchdog.cpp:
```diff
@@ -0,0 +1,57 @@
+#elif defined(__arm__)
+  if (WATCHDOG_PERIOD_MS <= 16000) {
+    WDT_Enable(WDT,
+               WDT_MR_WDRSTEN | WDT_MR_WDV(WATCHDOG_PERIOD_MS * 256 / 1000));
```
What is the point of the bitwise OR here? Does the watchdog timer `WDT_Enable` for ARM take two parameters?

mrrosen replied:
This function is just writing to the watchdog configuration register directly from the second argument, so it takes the bit fields directly; thus ORing together the enable bit and the bits for the period.

zmk5 replied:
Other than this question things seem good!

Commit:
alarms: compact alarm structures useing bitfields

Help minimize memory useage as alarm structures use a few fields
that are boolean values; bitfielding these and valid adjacent
fields can help reduce the amount of memory taken up by alarm
structures.

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

Commit:
main: move watchdog initialization to avoid resets during boot

For now, move the watchdog initialization as the motor hal takes
a while to home the motor as part of its initialization function.
Until this is refactored, the watchdog needs to only start running
after all the other things initialize.

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>
